File: ADT/dfs_bfs/127.py
from collections import deque


class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordSet = set(wordList)
        if endWord not in wordSet:
            return 0

        queue = deque([(beginWord, 1)])  # store (word, steps)

        while queue:
            word, steps = queue.popleft()

            if word == endWord:
                return steps

            for i in range(len(word)):
                for ch in "abcdefghijklmnopqrstuvwxyz":
                    next_word = word[:i] + ch + word[i+1:]

                    if next_word in wordSet:
                        wordSet.remove(next_word)  
                        queue.append((next_word, steps + 1))

        return 0

File: ADT/dfs_bfs/127.js
var ladderLength = function (beginWord, endWord, wordList) {
    const wordSet = new Set(wordList);
    if (!wordSet.has(endWord)) return 0;

    const queue = [[beginWord, 1]]; // store [word, steps]

    while (queue.length) {
        const [word, steps] = queue.shift();

        if (word === endWord) return steps;

        for (let i = 0; i < word.length; i++) {
            for (const ch of "abcdefghijklmnopqrstuvwxyz") {
                const next =
                    word.slice(0, i) + ch + word.slice(i + 1);

                if (wordSet.has(next)) {
                    wordSet.delete(next); // mark visited
                    queue.push([next, steps + 1]);
                }
            }
        }
    }
    return 0;
};
